File: ps/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;

    Node() {}

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node) {
            return NULL;
        }
        Node *copied = new Node(node->val, vector<Node*>());
        unordered_map<Node *,Node *> map;
        queue<Node *> que;

        map[node] = copied;
        que.push(node);
        while (!que.empty()) {
            Node *cur = que.front();
            que.pop();
            for (auto neighbor: cur->neighbors) {
                if (map.find(neighbor) == map.end()) {
                    que.push(neighbor);
                    map[neighbor] = new Node(neighbor->val, vector<Node*>());
                }
                map[cur]->neighbors.push_back(map[neighbor]);
            }
        }
        return copied;
    }
};

File: ps/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return
        deq = deque()
        copied = Node(node.val)
        mp = {node.val: copied}
        deq.append(node)
        while deq:
            node = deq.popleft()

            for neighbor in node.neighbors:
                if neighbor.val not in mp:

                    copy_neigh = Node(neighbor.val)
                    mp[neighbor.val] = copy_neigh
                    mp[node.val].neighbors.append(copy_neigh)

                    deq.append(neighbor)
                else:
                    mp[node.val].neighbors.append(mp[neighbor.val])

        return copied